server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java

```diff
@@ -31,36 +30,54 @@ public static void preProcess(SearchContext context) {
         if (context.aggregations() == null) {
             return;
         }
-        final Supplier<Collector> collectorSupplier;
+        final Supplier<AggregatorCollector> collectorSupplier;
         if (context.aggregations().isInSortOrderExecutionRequired()) {
-            executeInSortOrder(context, newBucketCollector(context));
-            collectorSupplier = () -> BucketCollector.NO_OP_COLLECTOR;
+            AggregatorCollector collector = newAggregatorCollector(context);
+            executeInSortOrder(context, collector.bucketCollector);
+            collectorSupplier = () -> new AggregatorCollector(collector.aggregators, BucketCollector.NO_OP_BUCKET_COLLECTOR);
         } else {
-            collectorSupplier = () -> newBucketCollector(context).asCollector();
+            collectorSupplier = () -> newAggregatorCollector(context);
         }
         context.aggregations().registerAggsCollectorManager(new CollectorManager<>() {
             @Override
-            public Collector newCollector() {
+            public AggregatorCollector newCollector() {
                 return collectorSupplier.get();
             }
 
             @Override
-            public Void reduce(Collection<Collector> collectors) {
-                // we cannot run post-collection method here because we need to do it after the optional timeout
-                // has been removed from the index searcher. Therefore, we delay this processing to the
-                // AggregationPhase#execute method.
+            public Void reduce(Collection<AggregatorCollector> collectors) {
+                if (context.queryResult().hasAggs()) {
+                    // no need to compute the aggs twice, they should be computed on a per context basis
+                    return null;
```

Comment:
Now that we moved this logic to the reduce, do we need to unset `context.aggregations(null);` and check for `if (context.queryResult().hasAggs()) {`? Or can reduce be invoked multiple times?

Reply:
I add it out of paranoia but I agree it might not be necessary